soc: atmel: sam/sam0: Add platform specific initialization

nandojve opened the pull request:
This replace pre kernel initialization by the platform specific initialization call. The platform specific init will configure at very beginning the clocks, flash wait states and cache.

~Supersedes #60425~

attie-argentum commented on soc/arm/atmel_sam0/common/soc_samd5x.c:
```diff
@@ -116,8 +117,14 @@ static int atmel_samd_init(void)
 	}
 
-	/* enable the Cortex M Cache Controller */
-	CMCC->CTRL.bit.CEN = 1;
+	/*
+	 * Force Cortex M Cache Controller disabled
+	 *
+	 * It is not clear if regular Cortex-M instructions can be used to
+	 * perform cache maintenance or this is a proprietary cache controller
+	 * that require special SoC support.
+	 */
+	CMCC->CTRL.bit.CEN = 0;
```
Please break this into a dedicated commit.

nandojve replied:
Done!

attie-argentum commented on soc/arm/atmel_sam/sam3x/Kconfig.series:
```diff
@@ -9,9 +9,10 @@ config SOC_SERIES_SAM3X
 	select ARM
 	select CPU_CORTEX_M3
-	select SOC_FAMILY_SAM
 	select CPU_HAS_ARM_MPU
 	select CPU_CORTEX_M_HAS_DWT
+	select PLATFORM_SPECIFIC_INIT
+	select SOC_FAMILY_SAM
```
Avoid unnecessary changes? (elsewhere too)

nandojve replied:
ditto

attie-argentum commented on soc/arm/atmel_sam/sam3x/soc.c:
```diff
@@ -192,15 +193,7 @@ static ALWAYS_INLINE void clock_init(void)
- * @return 0
- */
-static int atmel_sam3x_init(void)
+void z_arm_platform_init(void)
```
`z_arm_platform_init()` is a lot earlier than `PRE_KERNEL_1` - are you confident that this will work in all cases?

nandojve replied:
The z_arm_platform_init() seems to be the correct place to configure clocks, cache and any other specific details. This happen at reset which configure SoC ASAP. This speed up core initialization too.

The remaining PRE_KERNEL are initialization that require some kind OS resource.

nandojve committed:
soc: arm: atmel: Normalize Kconfig.series files

This update Kconfig.series files to normalize copyright date order from
lower to higher and reorder select entries.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: arm: atmel: Reorder board defconfig entries

This reorder <board>_defconfig entries in soc/board, internal and driver
definitions.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
soc: arm: sam: Reorder copyright by year ascending

Adjust copyright order by ascending year order.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
soc: atmel: Enable platform specific init

This replace pre kernel initialization by the platform specific
initialization call. The platform specific init will configure
at very beginning the clocks, flash wait states and cache.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
soc: atmel: sam: Fix cache management

The current platform initialization do not take in consideration cache
management for historic reasons. This fixes any miss configuration and
allow users to enable/disable caches at board definition. The default
value is cache disabled and the below examples are for SAMV71 which
have both I/D Cache available:

I Cache only:
CONFIG_CACHE_MANAGEMENT=y
CONFIG_DCACHE=n

D Cache only:
CONFIG_CACHE_MANAGEMENT=y
CONFIG_ICACHE=n

I/D Cache disabled:
CONFIG_ICACHE=n
CONFIG_DCACHE=n

I/D Cache Enabled:
CONFIG_CACHE_MANAGEMENT=y

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
soc: atmel: same5x: Disable cache

The sam0 CMCC configure Cortex-M cache controller. However, it is not
clear how the cache management should be performed. It is nor clear if
instructions like SCB_EnableICache can be used. In this case, if cache
management should be made only by CMCC it may require a dedicated
implementation.

Besides above, the CPU_CORTEX_M4 do not define cache by default which
can signal a bad configuration in tree since the SOC_SERIES_SAME54 do
not define which caches should be available.

This force cache controller disable to avoid issues.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented on boards/arm/sam_e70_xplained/sam_e70b_xplained_defconfig:
```diff
@@ -6,10 +6,17 @@ CONFIG_SOC_ATMEL_SAME70_EXT_MAINCK=y
 CONFIG_BUILD_OUTPUT_HEX=y
+
+# Enable HW stack protection
+CONFIG_HW_STACK_PROTECTION=y
```
Done!

nandojve commented on boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig:
```diff
@@ -6,14 +6,17 @@ CONFIG_SOC_ATMEL_SAME70_EXT_MAINCK=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_USART_SAM=y
-CONFIG_GPIO=y
```
I split those changes in other commits. I prefer that usually these configs follow same order to make it easy to compare between variants. I create 3 clusters to group related configs too.

attie-argentum reviewed (approved):
Thanks for the changes!
